palette/src/gradient.rs

```diff
@@ -142,7 +142,11 @@ impl<'a, C: Mix + Clone> Iterator for Take<'a, C> {
 
     fn next(&mut self) -> Option<C> {
         if self.from_head + self.from_end < self.len {
-            let i = self.from + (self.diff / cast(self.len)) * cast(self.from_head);
+            let i = if self.len > 1 {
+                self.from + (self.diff / cast(self.len - 1)) * cast(self.from_head)
+            } else {
+                self.from + (self.diff / cast(self.len)) * cast(self.from_head)
```

Comment:
I suppose it's possible to get here if `len`, `from_head` and `from_end` are all 0. That wouldn't be great. I think we can make special cases for 0 and 1 length iterators. 0 gives `None` on every call and `1` give the lower bound color (to be consistent with the old behavior) if `from_head` and `from_end` are both 0 (otherwise `None`). That may simplify the rest of the logic.